File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        dic = {
        '2':'abc',
        '3':'def',
        '4':'ghi',
        '5':'jkl',
        '6':'mno',
        '7':'pqrs',
        '8':'tuv',
        '9':'wxyz'
        }
        def combine(digits:str, cb:List[str])->List[str]:
            if not digits:
                return cb
            res = []
            if not cb:
                res = list(dic[digits[0]])
            else:
                for x in cb:
                    res += [x+y for y in dic[digits[0]]]
            return combine(digits[1:],res)
        return combine(digits,[])
            

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        Map<String,String> digM = new HashMap<>();
        digM.put("2","abc");
        digM.put("3","def");
        digM.put("4","ghi");
        digM.put("5","jkl");
        digM.put("6","mno");
        digM.put("7","pqrs");
        digM.put("8","tuv");
        digM.put("9","wxyz");

        List<String> res = new LinkedList<>();
        for(int i=0; i<digits.length(); i++){
            String letters = digM.get(digits.charAt(i)+"");
            if(res.size()==0){
                for(int j=0; j<letters.length(); j++){
                    res.add(letters.charAt(j)+"");
                }
            }
            else{
                int resL = res.size();
                List<String> resTemp = new LinkedList<>();
                for(int j=0; j<letters.length(); j++){
                    for(int k=0; k<resL; k++){
                        resTemp.add(res.get(k)+letters.charAt(j));
                    }
                }
                res = resTemp;
            }
        }

        return res;
    }
}